File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        
        def recur(node, cur_max, cur_min):
            # leaf node
            if not node:
                return cur_max - cur_min
            
            cur_max = max(cur_max, node.val)
            cur_min = min(cur_min, node.val)
            
            left = recur(node.left, cur_max, cur_min)
            right = recur(node.right, cur_max, cur_min)
        
            return max(left, right)
        
        return recur(root, root.val, root.val)
        

File: 1026-maximum-difference-between-node-and-ancestor/1026-maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int helper(TreeNode* node, int maxi, int mini){
        if (node == nullptr) {
            return maxi-mini;
        }
        maxi = max(maxi, node->val);
        mini = min(mini, node->val);
        
        int left = helper(node->left, maxi, mini);
        int right = helper(node->right, maxi, mini);
        
        return max(left, right);
    }
    
    int maxAncestorDiff(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        return helper(root, root->val, root->val);
        
    }
};